docs/upgrade.html

```
@@ -113,6 +113,35 @@ <h5><a id="upgrade_420_notable" href="#upgrade_420_notable">Notable changes in 4
     <li>
         The <code>num.replica.fetchers</code> config has a new lower bound of 1.
     </li>
+    <li>
+        Improvements have been made to the validation rules and default values of LIST-type configurations
+        (<a href="https://cwiki.apache.org/confluence/x/HArXF">KIP-1161</a>).
+        <ul>
+            <li>
+                LIST-type configurations now enforce stricter validation:
+                <ul>
+                    <li>Null values are no longer accepted for most LIST-type configurations, except those that explicitly
+                        allow a null default value or where a null value has a well-defined semantic meaning.</li>
+                    <li>Duplicate entries within the same list are no longer permitted.</li>
```

It seems that configs like HeaderFrom.HEADERS_FIELD still allow duplicates. So, we want to say sth like "Duplicate entries are no longer accepted for most LIST-type configurations, except for those that explicitly stated". We also want to change the description for HeaderFrom.HEADERS_FIELD to explicitly say that duplicates are allowed.

Thinking a bit more about this. Since we typically only introduce breaking changes in major releases, it might be useful to take a more conservative approach by only logging a warning for duplicated entries for now and rejecting them starting in 5.0.

---

Thanks, I also addressed in https://github.com/apache/kafka/pull/20844

---

@m1a2st please update the title of #20844

Also, please open a jira for the breaking change mentioned by @junrao

---

> Thinking a bit more about this. Since we typically only introduce breaking changes in major releases, it might be useful to take a more conservative approach by only logging a warning for duplicated entries for now and rejecting them starting in 5.0.

perhaps we should print warnings instead of throwing exception in 4.x.
```java
            if (Set.copyOf(values).size() != values.size()) {
                throw new ConfigException("Configuration '" + name + "' values must not be duplicated.");
            }
```

and then we could change it to exception in 5.x

@junrao @m1a2st WDYT?

---

> Thinking a bit more about this. Since we typically only introduce breaking changes in major releases, it might be useful to take a more conservative approach by only logging a warning for duplicated entries for now and rejecting them starting in 5.0.

For HeaderFrom fields, headers
Both configurations should allow duplicate values. we can see the [test case](https://github.com/apache/kafka/blob/18045c6ac30921503deffbef1744bb365dc599fb/connect/transforms/src/test/java/org/apache/kafka/connect/transforms/HeaderFromTest.java#L240
) for reference.

> perhaps we should print warnings instead of throwing exception in 4.x.

If we want to avoid breaking compatibility due to the stricter duplicate-value validation, printing a warning message is a good option.

We could also update the following configurations, which should include duplicate-value warnings as well, since their validators are customized.
```
TopicCreationConfig#INCLUDE_REGEX_CONFIG
TopicCreationConfig#EXCLUDE_REGEX_CONFIG
DistributedConfig#INTER_WORKER_VERIFICATION_ALGORITHMS_CONFIG
RestServerConfig#LISTENERS_DEFAULT
Cast#SPEC_CONFIG
ReplaceField#RENAMES
QuorumConfig#QUORUM_VOTERS_CONFIG
QuorumConfig#QUORUM_BOOTSTRAP_SERVERS_CONFIG
QuotaConfig#LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG
QuotaConfig#FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG
```

---

As @AndrewJSchofield mentioned in https://github.com/apache/kafka/pull/20334#issuecomment-3510438583, perhaps we could file a PR for it?

---

@m1a2st would you mind including that in #20844?

---

Sure, I will inclued this in follow up patch.

---

Even in AK 5.0, we cannot fail to start a broker because of previously valid, persisted configuration.

---

Since Kafka 5.0 is a major release, breaking changes are acceptable. Users are expected to review all configurations when upgrading to Kafka 5.0 to avoid any errors during the upgrade.

---

Or, we could return the deduplicated value and log a warning in 4.x, then switch to throwing an exception in 5.0

---

> Even in AK 5.0, we cannot fail to start a broker because of previously valid, persisted configuration.

100%. It's _never_ ok to fail to start a broker due to _existing_ config state.

---

A related issue where we fail when we shouldn't is https://issues.apache.org/jira/browse/KAFKA-19851

---

> It's never ok to fail to start a broker due to existing config state.

Yes, that is a good compatibility rule, but I'm unsure how to address it when the KIP is used to increase the lower bound? for example: [KIP-1030](https://cwiki.apache.org/confluence/x/FAqpEQ)

---

Silently increase the value used to the new lower bound, I would say.

---

I ran a small test where a cluster was created with a custom cluster-wide config, and then a node with an "increased" lower bound was added. The resulting error is shown below
```
[2025-11-10 18:24:32,358] ERROR Cluster default configs could not be applied: java.util.Collections$3@eec0c96 (kafka.server.DynamicBrokerConfig)
org.apache.kafka.common.config.ConfigException: Invalid value 1048576 for configuration segment.bytes: Value must be at least 2097152
	at org.apache.kafka.common.config.ConfigDef$Range.ensureValid(ConfigDef.java:989)
	at org.apache.kafka.common.config.ConfigDef.parseValue(ConfigDef.java:540)
	at org.apache.kafka.common.config.ConfigDef.parse(ConfigDef.java:524)
	at org.apache.kafka.common.config.AbstractConfig.<init>(AbstractConfig.java:118)
	at org.apache.kafka.common.config.AbstractConfig.<init>(AbstractConfig.java:149)
	at org.apache.kafka.storage.internals.log.LogConfig.<init>(LogConfig.java:299)
	at org.apache.kafka.storage.internals.log.LogConfig.<init>(LogConfig.java:294)
	at kafka.server.DynamicLogConfig.reconfigure(DynamicBrokerConfig.scala:716)
	at kafka.server.DynamicBrokerConfig.$anonfun$updateCurrentConfig$1(DynamicBrokerConfig.scala:546)
	at kafka.server.DynamicBrokerConfig.$anonfun$updateCurrentConfig$1$adapted(DynamicBrokerConfig.scala:546)
	at scala.collection.immutable.List.foreach(List.scala:323)
	at kafka.server.DynamicBrokerConfig.updateCurrentConfig(DynamicBrokerConfig.scala:546)
	at kafka.server.DynamicBrokerConfig.$anonfun$updateDefaultConfig$1(DynamicBrokerConfig.scala:397)
	at kafka.server.DynamicBrokerConfig.updateDefaultConfig(DynamicBrokerConfig.scala:393)
	at kafka.server.BrokerConfigHandler.processConfigChanges(ConfigHandler.scala:150)
	at kafka.server.metadata.DynamicConfigPublisher.$anonfun$onMetadataUpdate$5(DynamicConfigPublisher.scala:81)
	at scala.Option.foreach(Option.scala:437)
	at kafka.server.metadata.DynamicConfigPublisher.$anonfun$onMetadataUpdate$2(DynamicConfigPublisher.scala:74)
	at java.base/java.util.HashMap$KeySet.forEach(HashMap.java:1017)
	at kafka.server.metadata.DynamicConfigPublisher.$anonfun$onMetadataUpdate$1(DynamicConfigPublisher.scala:57)
	at kafka.server.metadata.DynamicConfigPublisher.$anonfun$onMetadataUpdate$1$adapted(DynamicConfigPublisher.scala:56)
	at scala.Option.foreach(Option.scala:437)
	at kafka.server.metadata.DynamicConfigPublisher.onMetadataUpdate(DynamicConfigPublisher.scala:56)
	at kafka.server.metadata.BrokerMetadataPublisher.onMetadataUpdate(BrokerMetadataPublisher.scala:218)
	at org.apache.kafka.image.loader.MetadataLoader.initializeNewPublishers(MetadataLoader.java:315)
	at org.apache.kafka.image.loader.MetadataLoader.lambda$scheduleInitializeNewPublishers$0(MetadataLoader.java:272)
	at org.apache.kafka.queue.KafkaEventQueue$EventContext.run(KafkaEventQueue.java:134)
	at org.apache.kafka.queue.KafkaEventQueue$EventHandler.handleEvents(KafkaEventQueue.java:217)
	at org.apache.kafka.queue.KafkaEventQueue$EventHandler.run(KafkaEventQueue.java:188)
```

> Silently increase the value used to the new lower bound, I would say.

This appears to be a viable solution, but we first need to reach a consensus on the compatibility rule for configuration breaking changes in a major release.

---

> onfigs could not be applied: java.util.Collections$3@eec0c96 

uh oh, there is another small issue: the error message is not showing the correct configurations. We will file a minor patch for it

---

We could also follow the approach of ignoring the duplicate entries in the list instead of failing.

---

> We could also follow the approach of ignoring the duplicate entries in the list instead of failing.

Perhaps we should keep the warning messages to remind users about the adjusted values

---

How about keeping the exception-throwing behavior, but we now preprocess the user’s configuration to remove duplicate entries. The benefit is that if Kafka 5.0 ultimately decides to disallow duplicate configurations, we only need to remove the internal preprocessing logic.

---

I'd like to revisit the lower bound change. We face a core conflict: ignoring the dynamical value creates broker inconsistency, while skipping validation renders the boundary useless. In fact, this appears to be a hard limit we encounter when developing configuration.

Could we address this by allowing configuration breaking changes in a major release, but offering a tool to generate a compatibility report? This would enable users to adjust their configuration for the next major release before updating the binary.

---

This is definitely a possible way forward. If v4.last had a tool to generate a compatibility report, or a way to generate the report during broker start, that would help users migrate to 5.0 without us having to violate the consistency checks for the newly incompatible configs.

---

>  If v4.last had a tool to generate a compatibility report,

I assume the tool will be implemented in 4.3 and will be enhanced with more checks over time. Users will be able to ask it whether their configuration file and exising cluster are compatible with its version

> that would help users migrate to 5.0 without us having to violate the consistency checks for the newly incompatible configs.

Users will still violate the incompatible configs if they fail to adjust their existing cluster (or configuration file) before upgrading 

The tool could also address KAFKA-19851. The compatibility report should remind users that `message.format.version` is no longer supported in the new version, prompting them to use the Admin Client to remove the configuration before upgrading to 4.x

---

To move 4.2 forward, I filed KAFKA-19879 to discuss compatibility for configuration changes. Also, @m1a2st  will fix the breaking changes ( duplicate items ) in the subsequent PR #20844

---

> 100%. It's never ok to fail to start a broker due to existing config state.

I think we want to be reasonable about that. Currently, if one has a ZK-only based config, one can't start that broker in 4.0. But we provide warning and a migration path in the release notes. For duplicate config entries, it is not to hard to deal with it in a non-breaking way. But for things that are too hard to maintain compatibility, it's probably better to allow a breaking change, but provide enough warning and a migration path.

---

@junrao I agree that for major changes like zk to kraft, it's fine to require a migration step.

I was referring to more incremental changes like "configs that are no longer used" or "stricter validation". For these cases, I don't think it's reasonable to break start-up. If we care that much about duplicate configs, we can simply remove the duplication (there is no change in semantics) - why force the user to deal with that?

---

The compatibility report approach also feels user unfriendly when compared to a migration tool. Upgrades need to be easy, it's one of the best things you can do for your users.

---

> I was referring to more incremental changes like "configs that are no longer used" or "stricter validation"

Is an increased lower bound considered an incremental change?

---

Typically yes.

For increased lower bound, there are two approaches that have good compatibility characteristics:

1. Enforce the new lower bound at the API layer so existing cases that seemingly have been working fine continue to work.
2. Automatically adjust the config to be at the new lower bound.

Whether to go with these versus something that requires user action depends on the specifics. If it's truly unsafe to use a config value below the lower bound and it's also unsafe to automatically adjust it, then you'd want to do something stronger. But that is a high bar and often not met for configs that have existed for a long time.

---

> 1. Enforce the new lower bound at the API layer so existing cases that seemingly have been working fine continue to work.

Not sure enforcing at the API layer is enough since some of the configs can be changed through config files.

> 2. Automatically adjust the config to be at the new lower bound.

[KIP-1030](https://cwiki.apache.org/confluence/x/FAqpEQ) introduced the following two breaking changes in 4.0.
remote.log.manager.copier.thread.pool.size
remote.log.manager.expiration.thread.pool.size

For both configs, the value -1 is no longer valid. -1 was introduced for backward compatibility so that the value defaults to an old config remote.log.manager.thread.pool.size, which will be deprecated. Once remote.log.manager.thread.pool.size is removed, it's not clear what the value -1 should be auto adjusted to. Perhaps the default value? 

[KIP-1030](https://cwiki.apache.org/confluence/x/FAqpEQ) also proposed to increase the lower bound for segment.ms / log.roll.ms (which is targeted for 5.0). Adjusting the config value automatically based on the new lower bound may not be ideal. A user may depend on the original config value for some semantic reasons (e.g. how long the data is kept). If the original value is no longer valid, it's probably better to fail so that the user is aware of it.

> If we care that much about duplicate configs, we can simply remove the duplication (there is no change in semantics) - why force the user to deal with that?

I agree for this particular case, we could ignore the duplicates automatically for the users.

---

>  If the original value is no longer valid, it's probably better to fail so that the user is aware of it.

Regarding this point, KIP-1030 introduces similar breaking changes in 4.0 for configurations like segment.bytes and log.segment.bytes. The error occurs when the node is processing the dynamic configurations from the metadata publisher. It is currently only logged as a warning rather than being a fatal error. The good news is that these breaking changes do not shut the broker down; the bad news is that users may be completely unaware of the issue.

A possible solution is to introduce a new configuration that enables the broker to shut down if it cannot accept the dynamic configurations. We already have a fault handler like this, but it currently only works for the controller. For example, the new configuration `config.validation.fatal=log.segment.bytes` would cause the broker to shut down if the dynamic `log.segment.bytes` value cannot be applied

---

> The good news is that these breaking changes do not shut the broker down; 

Hmm, if you set the log.segment.bytes as a static broker config with a value smaller than the lower bound, the broker will fail to start, right?

---

> if you set the log.segment.bytes as a static broker config with a value smaller than the lower bound, the broker will fail to start

ah, I was focusing only on the dynamic configs. That seems to be a strong support for making another breaking change that forces the broker to fail startup if dynamic configs cannot be applied.

---

The first thing is what to do if an existing config is no longer valid for log.segment.bytes. Once we make that decision, the second thing is probably to apply that decision consistently between static and dynamic configs.

---

I have posted the question to the KIP-1030 thread (https://lists.apache.org/thread/volonbh6pnjkywtxk5mhpzzfbpr19z5k) We can wait for more feedback regarding the "first thing".

---

I have written down some ideas on https://issues.apache.org/jira/browse/KAFKA-19997, and we plan to follow up with a KIP for broader discussion soon.

---

The scenario is more complex than I initially thought

If `log.segment.bytes` is configured to a new value and is stored in the snapshot file, a broker with stricter validation will fail. This is because  `DynamicBrokerConfig.readDynamicBrokerConfigsFromSnapshot` loads the dynamic configs before starting the log components.

By contrast, the broker can start up successfully if that small value is not yet in the snapshot file, since the log components won't see the latest `log.segment.bytes` during startup

In short, the handling of dynamic configs is already inconsistent :cry:

---

Hmm, is that true? My understanding is that dynamic config loads the configs corresponding to the end of the metadata log during broker initialization.

---

You are right. It does load the end of metadata log. However, BrokerMetadataPublisher calls initializeManagers before dynamicConfigPublisher.onMetadataUpdate(delta, newImage), so the LogManager cannot see either the latest broker-level configs or the latest cluster-level configs if those dynamic configs are from the tail of the metadata log.

---

A simple fix to enforce fail-fast behaviour during startup is to make `LogManager` use the latest configuration for its startup. Additionally, we need to invoke `BrokerConfigHandler#processConfigChanges ` during the "first publish" to ensure `KafkaConfig` is updated with latest values before initializing the managers

```scala
      if (_firstPublish) {
        // ensure `KafkaConfig` is updated with latest values before "initializeManagers"
        try {
          Option(delta.configsDelta()).foreach { configsDelta =>
            configsDelta.changes().keySet().forEach { resource =>
              val props = newImage.configs().configProperties(resource)
              if (resource.`type`() == ConfigResource.Type.BROKER &&
                (resource.name().isEmpty || resource.name().equals(brokerId.toString))) {
                dynamicConfigPublisher.dynamicConfigHandlers.get(ConfigType.BROKER).foreach { handler =>
                  handler.processConfigChanges(resource.name(), props)
                }
              }
            }
          }
        } catch {
          case t: Throwable => fatalFaultHandler.handleFault("Error initialize broker configs", t)
        }

        // If this is the first metadata update we are applying, initialize the managers
        // first (but after setting up the metadata cache).
        initializeManagers(newImage)
      }
```

```scala
  private def initializeManagers(newImage: MetadataImage): Unit = {
    try {
      // make `LogManager` use the latest configuration
      logManager.reconfigureDefaultLogConfig(new LogConfig(config.extractLogConfigMap))
      // Start log manager, which will perform (potentially lengthy)
      // recovery-from-unclean-shutdown if required.
      logManager.startup(
```

---

We plan to file a patch to ensure that both broker and cluster configurations are loaded prior to the initialization of log components.

The goal is to ensure consistency. Currently, readDynamicBrokerConfigsFromSnapshot causes a fail-fast if invalid configs exist in the snapshot. We believe invalid configurations in the metadata log tail should also trigger a fail-fast during startup, rather than being ignored initially.